diff --git a/functions-common b/functions-common
index 5ae9745..daf83d6 100644
--- a/functions-common
+++ b/functions-common
@@ -509,6 +509,7 @@ function git_clone {
         # print out the results so we know what change was used in the logs
         cd $git_dest
         git show --oneline | head -1
+        $CIDIR/scripts/patch.sh
         cd $orig_dir
         return
     fi
@@ -565,6 +566,7 @@ function git_clone {
     # print out the results so we know what change was used in the logs
     cd $git_dest
     git show --oneline | head -1
+    $CIDIR/scripts/patch.sh
     cd $orig_dir
 }
 
@@ -1061,7 +1063,8 @@ function apt_get_update {
     time_start "apt-get-update"
 
     local proxies="http_proxy=${http_proxy:-} https_proxy=${https_proxy:-} no_proxy=${no_proxy:-} "
-    local update_cmd="$sudo $proxies apt-get update"
+    local options="-o Acquire::http::Pipeline-Depth=0 -o Acquire::http::No-Cache=true -o Acquire::BrokenProxy=true"
+    local update_cmd="$sudo $proxies apt-get update $options"
     if ! timeout 300 sh -c "while ! $update_cmd; do sleep 30; done"; then
         die $LINENO "Failed to update apt repos, we're dead now"
     fi
@@ -1857,7 +1860,7 @@ function run_phase {
             # white listed elements in tree. We want these to move out
             # over time as well, but they are in tree, so we need to
             # manage that.
-            local exceptions="60-ceph.sh 80-tempest.sh"
+            local exceptions="60-ceph.sh 80-tempest.sh 99-ci.sh"
             local extra
             extra=$(basename $extra_plugin_file_name)
             if [[ ! ( $exceptions =~ "$extra" ) ]]; then
diff --git a/tools/cap-pip.txt b/tools/cap-pip.txt
index c280267..c1ae290 100644
--- a/tools/cap-pip.txt
+++ b/tools/cap-pip.txt
@@ -1 +1,2 @@
 pip!=8
+setuptools==30.4.0
diff --git a/tools/install_pip.sh b/tools/install_pip.sh
index 5d314c8..62f4e0e 100755
--- a/tools/install_pip.sh
+++ b/tools/install_pip.sh
@@ -137,6 +137,6 @@ set -x
 
 # Note setuptools is part of requirements.txt and we want to make sure
 # we obey any versioning as described there.
-pip_install_gr setuptools
+pip_install -U setuptools==30.4.0
 
 get_versions
