diff --git a/cinder/exception.py b/cinder/exception.py
index 553f06e..bfcd5cd 100644
--- a/cinder/exception.py
+++ b/cinder/exception.py
@@ -720,3 +720,7 @@ class BrocadeZoningCliException(CinderException):
 
 class NetAppDriverException(VolumeDriverException):
     message = _("NetApp Cinder Driver exception.")
+
+
+class NexentaException(NotFound):
+    message = _("Nexenta Cinder Driver exception.")
diff --git a/etc/cinder/rootwrap.d/volume.filters b/etc/cinder/rootwrap.d/volume.filters
index 5b57404..caf8755 100644
--- a/etc/cinder/rootwrap.d/volume.filters
+++ b/etc/cinder/rootwrap.d/volume.filters
@@ -74,6 +74,7 @@ netapp_nfs_find: RegExpFilter, find, root, find, ^[/]*([^/\0]+(/+)?)*$, -maxdept
 
 # cinder/volume/drivers/glusterfs.py
 chgrp: CommandFilter, chgrp, root
+umount: CommandFilter, umount, root
 
 # cinder/volumes/drivers/hds/hds.py:
 hus-cmd: CommandFilter, hus-cmd, root
diff --git a/requirements.txt b/requirements.txt
index c5d8e94..26b6222 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -5,6 +5,7 @@ argparse
 Babel>=1.3,<=1.3
 eventlet>=0.13.0,<0.16.0
 greenlet>=0.3.2,<=0.4.5
+ipaddress==1.0.22
 iso8601>=0.1.9,<=0.1.10
 kombu>=2.5.0,<=3.0.7
 lxml>=2.3,<=3.4.2
diff --git a/tox.ini b/tox.ini
index 42b139d..7474ee6 100644
--- a/tox.ini
+++ b/tox.ini
@@ -9,10 +9,11 @@ envlist = py26,py27,py33,pep8
 setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
 usedevelop = True
-install_command = pip install {opts} {packages}
+install_command = pip install {opts} -c{toxinidir}/../requirements/upper-constraints.txt {packages}
 
 deps = -r{toxinidir}/requirements.txt
        -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
   python setup.py testr --slowest --testr-args='--concurrency 1 {posargs}'
 
@@ -26,6 +27,7 @@ commands =
 
 [testenv:pylint]
 deps = -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        pylint==0.26.0
 commands = bash tools/lintstack.sh
 
