diff --git a/cinder/api/openstack/wsgi.py b/cinder/api/openstack/wsgi.py
index 1fc085951..e6d9adee7 100644
--- a/cinder/api/openstack/wsgi.py
+++ b/cinder/api/openstack/wsgi.py
@@ -17,6 +17,7 @@
 import inspect
 import math
 import time
+import sys
 from xml.dom import minidom
 from xml.parsers import expat
 
@@ -37,6 +38,9 @@ XML_NS_V1 = 'http://docs.openstack.org/api/openstack-block-storage/1.0/content'
 XML_NS_V2 = 'http://docs.openstack.org/api/openstack-block-storage/2.0/content'
 XML_NS_ATOM = 'http://www.w3.org/2005/Atom'
 
+reload(sys)
+sys.setdefaultencoding('utf-8')
+
 LOG = logging.getLogger(__name__)
 
 SUPPORTED_CONTENT_TYPES = (
diff --git a/cinder/exception.py b/cinder/exception.py
index 60c571c11..413910cb9 100644
--- a/cinder/exception.py
+++ b/cinder/exception.py
@@ -457,6 +457,10 @@ class FailedCmdWithDump(VolumeDriverException):
     message = _("Operation failed with status=%(status)s. Full dump: %(data)s")
 
 
+class NexentaException(VolumeDriverException):
+    message = _("Operation failed with status=%(status)s. Full dump: %(data)s")
+
+
 class GlanceMetadataExists(Invalid):
     message = _("Glance metadata cannot be updated, key %(key)s"
                 " exists for volume id %(volume_id)s")
