diff --git a/cinder/exception.py b/cinder/exception.py
index 60c571c..413910c 100644
--- a/cinder/exception.py
+++ b/cinder/exception.py
@@ -457,6 +457,10 @@ class FailedCmdWithDump(VolumeDriverException):
     message = _("Operation failed with status=%(status)s. Full dump: %(data)s")
 
 
+class NexentaException(VolumeDriverException):
+    message = _("Operation failed with status=%(status)s. Full dump: %(data)s")
+
+
 class GlanceMetadataExists(Invalid):
     message = _("Glance metadata cannot be updated, key %(key)s"
                 " exists for volume id %(volume_id)s")
