diff --git a/cinder/exception.py b/cinder/exception.py
index 014ab3d..fc8d888 100644
--- a/cinder/exception.py
+++ b/cinder/exception.py
@@ -497,6 +497,10 @@ class FailedCmdWithDump(VolumeDriverException):
     message = _("Operation failed with status=%(status)s. Full dump: %(data)s")
 
 
+class NexentaException(VolumeDriverException):
+    message = _("Operation failed with status=%(status)s. Full dump: %(data)s")
+
+
 class InvalidConnectorException(VolumeDriverException):
     message = _("Connector doesn't have required information: %(missing)s")
 
diff --git a/tox.ini b/tox.ini
index 3060c8a..c22c037 100644
--- a/tox.ini
+++ b/tox.ini
@@ -9,10 +9,11 @@ envlist = py27,pep8
 setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
 usedevelop = True
-install_command = pip install {opts} {packages}
+install_command = pip install -c{toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 
 deps = -r{toxinidir}/requirements.txt
        -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
   bash tools/pretty_tox.sh '{posargs}'
 whitelist_externals = bash
@@ -28,6 +29,7 @@ commands =
 
 [testenv:pylint]
 deps = -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        pylint==0.26.0
 commands = bash tools/lintstack.sh
 
