diff --git a/cinder/api/openstack/wsgi.py b/cinder/api/openstack/wsgi.py
index e5f0508..9b22608 100644
--- a/cinder/api/openstack/wsgi.py
+++ b/cinder/api/openstack/wsgi.py
@@ -47,6 +47,11 @@ from cinder.i18n import _
 from cinder import utils
 from cinder.wsgi import common as wsgi
 
+if six.PY2:
+    import sys
+    from imp import reload
+    reload(sys)
+    sys.setdefaultencoding('utf-8')
 
 LOG = logging.getLogger(__name__)
 
diff --git a/cinder/api/v3/volumes.py b/cinder/api/v3/volumes.py
index e951817..e93ae82 100644
--- a/cinder/api/v3/volumes.py
+++ b/cinder/api/v3/volumes.py
@@ -194,12 +194,6 @@ class VolumeController(volumes_v2.VolumeController):
                     "the latest one of volume %(v_id)s.")
             raise exc.HTTPBadRequest(explanation=msg % {'s_id': snapshot_id,
                                                         'v_id': volume.id})
-        if volume.size != l_snap.volume_size:
-            msg = _("Can't revert volume %(v_id)s to its latest snapshot "
-                    "%(s_id)s. The volume size must be equal to the snapshot "
-                    "size.")
-            raise exc.HTTPBadRequest(explanation=msg % {'s_id': snapshot_id,
-                                                        'v_id': volume.id})
         try:
             msg = 'Reverting volume %(v_id)s to snapshot %(s_id)s.'
             LOG.info(msg, {'v_id': volume.id,
diff --git a/tox.ini b/tox.ini
index 4b0141e..8c16daf 100644
--- a/tox.ini
+++ b/tox.ini
@@ -14,7 +14,7 @@ setenv = VIRTUAL_ENV={envdir}
 usedevelop = True
 install_command = pip install {opts} {packages}
 
-deps = -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+deps = -c{toxinidir}/../requirements/upper-constraints.txt
         -r{toxinidir}/test-requirements.txt
         -r{toxinidir}/requirements.txt
 
@@ -133,7 +133,7 @@ commands = {posargs}
 [testenv:docs]
 basepython = python3
 deps =
-  -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+  -c{toxinidir}/../requirements/upper-constraints.txt
   -r{toxinidir}/doc/requirements.txt
 commands =
   rm -fr doc/source/contributor/api/
