diff --git a/tox.ini b/tox.ini
index 3c435b6..91cc5e9 100644
--- a/tox.ini
+++ b/tox.ini
@@ -9,7 +9,7 @@ envlist = py27,pep8
 setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
 usedevelop = True
-install_command = pip install -c{env:UPPER_CONSTRAINTS_FILE:https://git.openstack.org/cgit/openstack/requirements/plain/upper-constraints.txt?h=stable/mitaka} {opts} {packages}
+install_command = pip install -c{toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 
 # TODO(mriedem): Move oslo.versionedobjects[fixtures] to test-requirements.txt
 #                after I937823ffeb95725f0b55e298ebee1857d6482883 lands.
