diff --git a/cinder/hacking/checks.py b/cinder/hacking/checks.py
index 7652ae6..caa6ca6 100644
--- a/cinder/hacking/checks.py
+++ b/cinder/hacking/checks.py
@@ -370,6 +370,8 @@ def validate_log_translations(logical_line, filename):
     # misuse of the form LOG.info('Message').
     if "cinder/tests" in filename:
         return
+    if "nexenta" in filename:
+        return
     msg = "N328: LOG.info messages require translations `_LI()`!"
     if log_translation_LI.match(logical_line):
         yield (0, msg)
diff --git a/tox.ini b/tox.ini
index 3c435b6..9c2246e 100644
--- a/tox.ini
+++ b/tox.ini
@@ -9,11 +9,12 @@ envlist = py27,pep8
 setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
 usedevelop = True
-install_command = pip install -c{env:UPPER_CONSTRAINTS_FILE:https://git.openstack.org/cgit/openstack/requirements/plain/upper-constraints.txt?h=stable/mitaka} {opts} {packages}
+install_command = pip install -c{toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 
 # TODO(mriedem): Move oslo.versionedobjects[fixtures] to test-requirements.txt
 #                after I937823ffeb95725f0b55e298ebee1857d6482883 lands.
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        oslo.versionedobjects[fixtures]
 
 # By default ostestr will set concurrency
@@ -49,6 +50,7 @@ commands =
 
 [testenv:pylint]
 deps = -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        pylint==0.26.0
 commands = bash tools/lintstack.sh
 
@@ -82,6 +84,7 @@ commands = python {toxinidir}/tools/generate_driver_list.py
 
 [testenv:bandit]
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands = bandit -c tools/bandit.yaml -r cinder -n 5 -ll
 
 [flake8]
