diff --git a/tox.ini b/tox.ini
index 563e0cf..e18fbd1 100644
--- a/tox.ini
+++ b/tox.ini
@@ -9,11 +9,12 @@ envlist = py34,py27,compliance,pep8
 setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
 usedevelop = True
-install_command = pip install -c{env:UPPER_CONSTRAINTS_FILE:https://git.openstack.org/cgit/openstack/requirements/plain/upper-constraints.txt?h=stable/newton} {opts} {packages}
+install_command = pip install -c{toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 
 # TODO(mriedem): Move oslo.versionedobjects[fixtures] to test-requirements.txt
 #                after I937823ffeb95725f0b55e298ebee1857d6482883 lands.
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        oslo.versionedobjects[fixtures]
 
 # By default ostestr will set concurrency
@@ -34,6 +35,7 @@ passenv = *_proxy *_PROXY
 # the API Ref to developer.openstack.org.
 whitelist_externals = rm
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
   rm -rf api-ref/build
   sphinx-build -W -b html -d api-ref/build/doctrees api-ref/source api-ref/build/html/
@@ -71,6 +73,7 @@ commands =
 
 [testenv:pylint]
 deps = -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        pylint==0.26.0
 commands = bash tools/lintstack.sh
 
@@ -108,6 +111,7 @@ commands = python {toxinidir}/tools/generate_driver_list.py
 
 [testenv:bandit]
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands = bandit -r cinder -n5 -x tests -ll
 
 [flake8]
