diff --git a/cinder/api/openstack/wsgi.py b/cinder/api/openstack/wsgi.py
index 8a19d18..9e5c187 100644
--- a/cinder/api/openstack/wsgi.py
+++ b/cinder/api/openstack/wsgi.py
@@ -18,6 +18,7 @@ import functools
 import inspect
 import math
 import time
+import sys
 
 from oslo_log import log as logging
 from oslo_serialization import jsonutils
@@ -37,6 +38,8 @@ from cinder import policy
 from cinder import utils
 from cinder.wsgi import common as wsgi
 
+reload(sys)
+sys.setdefaultencoding('utf-8')
 
 LOG = logging.getLogger(__name__)
 
diff --git a/cinder/volume/manager.py b/cinder/volume/manager.py
index 490b406..7808dea 100644
--- a/cinder/volume/manager.py
+++ b/cinder/volume/manager.py
@@ -38,6 +38,7 @@ intact.
 
 import requests
 import time
+import sys
 
 from oslo_config import cfg
 from oslo_log import log as logging
@@ -80,6 +81,9 @@ from cinder.volume import rpcapi as volume_rpcapi
 from cinder.volume import utils as vol_utils
 from cinder.volume import volume_types
 
+reload(sys)
+sys.setdefaultencoding('utf-8')
+
 LOG = logging.getLogger(__name__)
 
 QUOTAS = quota.QUOTAS
