diff --git a/cinder/api/openstack/wsgi.py b/cinder/api/openstack/wsgi.py
index 0e6d04b..e33de65 100644
--- a/cinder/api/openstack/wsgi.py
+++ b/cinder/api/openstack/wsgi.py
@@ -37,6 +37,11 @@ from cinder import policy
 from cinder import utils
 from cinder.wsgi import common as wsgi
 
+if six.PY2:
+    from imp import reload
+    import sys
+    reload(sys)
+    sys.setdefaultencoding('utf-8')
 
 LOG = logging.getLogger(__name__)
 
diff --git a/cinder/hacking/checks.py b/cinder/hacking/checks.py
index 804b54c..d830b27 100644
--- a/cinder/hacking/checks.py
+++ b/cinder/hacking/checks.py
@@ -372,6 +372,8 @@ def validate_log_translations(logical_line, filename):
     # misuse of the form LOG.info('Message').
     if "cinder/tests" in filename:
         return
+    if "nexenta" in filename:
+        return
     msg = "N328: LOG.info messages require translations `_LI()`!"
     if log_translation_LI.match(logical_line):
         yield (0, msg)
diff --git a/cinder/tests/tempest/api/volume/base.py b/cinder/tests/tempest/api/volume/base.py
index 50ee60a..01f0e16 100644
--- a/cinder/tests/tempest/api/volume/base.py
+++ b/cinder/tests/tempest/api/volume/base.py
@@ -106,11 +106,6 @@ class BaseVolumeTest(api_version_utils.BaseMicroversionTest,
             kwargs['name'] = name
 
         volume = cls.volumes_client.create_volume(**kwargs)['volume']
-        cls.addClassResourceCleanup(
-            cls.volumes_client.wait_for_resource_deletion, volume['id'])
-        cls.addClassResourceCleanup(test_utils.call_and_ignore_notfound_exc,
-                                    cls.volumes_client.delete_volume,
-                                    volume['id'])
         waiters.wait_for_volume_resource_status(cls.volumes_client,
                                                 volume['id'], wait_until)
         return volume
diff --git a/cinder/tests/tempest/api/volume/test_consistencygroups.py b/cinder/tests/tempest/api/volume/test_consistencygroups.py
index d6c7de5..b898876 100644
--- a/cinder/tests/tempest/api/volume/test_consistencygroups.py
+++ b/cinder/tests/tempest/api/volume/test_consistencygroups.py
@@ -14,18 +14,18 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+from tempest.api.volume import base
 from tempest.common import waiters
 from tempest import config
 from tempest.lib.common.utils import data_utils
 from tempest import test
 
-from cinder.tests.tempest.api.volume import base
 from cinder.tests.tempest import cinder_clients
 
 CONF = config.CONF
 
 
-class ConsistencyGroupsV2Test(base.BaseVolumeTest):
+class ConsistencyGroupsV2Test(base.BaseVolumeAdminTest):
     @classmethod
     def setup_clients(cls):
         cls._api_version = 2
diff --git a/cinder/tests/tempest/api/volume/test_volume_backup.py b/cinder/tests/tempest/api/volume/test_volume_backup.py
index 717c3b3..9e8ea4e 100644
--- a/cinder/tests/tempest/api/volume/test_volume_backup.py
+++ b/cinder/tests/tempest/api/volume/test_volume_backup.py
@@ -62,6 +62,7 @@ class VolumesBackupsTest(base.BaseVolumeTest):
         """Test backup create and restore to an existing volume."""
         # Create volume
         src_vol = self.create_volume()
+        self.addCleanup(self.volumes_client.delete_volume, src_vol['id'])
         # Create backup
         backup = self.create_backup(volume_id=src_vol['id'])
         # Restore to existing volume
@@ -84,6 +85,7 @@ class VolumesBackupsTest(base.BaseVolumeTest):
         volume = self.create_volume(size=CONF.volume.volume_size,
                                     imageRef=CONF.compute.image_ref)
 
+        self.addCleanup(self.volumes_client.delete_volume, volume['id'])
         # Create backup
         self.create_backup(volume_id=volume['id'])
         # Create a server
diff --git a/cinder/tests/tempest/api/volume/test_volume_unicode.py b/cinder/tests/tempest/api/volume/test_volume_unicode.py
index db280d3..1ab7c10 100644
--- a/cinder/tests/tempest/api/volume/test_volume_unicode.py
+++ b/cinder/tests/tempest/api/volume/test_volume_unicode.py
@@ -17,7 +17,6 @@
 from tempest.common import waiters
 from tempest import config
 from tempest.lib.common.utils import data_utils
-from tempest.lib.common.utils import test_utils
 
 from cinder.tests.tempest.api.volume import base
 
@@ -44,11 +43,6 @@ class CinderUnicodeTest(base.BaseVolumeTest):
         kwargs['size'] = CONF.volume.volume_size
 
         volume = cls.volumes_client.create_volume(**kwargs)['volume']
-        cls.addClassResourceCleanup(
-            cls.volumes_client.wait_for_resource_deletion, volume['id'])
-        cls.addClassResourceCleanup(test_utils.call_and_ignore_notfound_exc,
-                                    cls.volumes_client.delete_volume,
-                                    volume['id'])
         waiters.wait_for_volume_resource_status(cls.volumes_client,
                                                 volume['id'],
                                                 'available')
@@ -62,3 +56,4 @@ class CinderUnicodeTest(base.BaseVolumeTest):
         fetched_volume = result['volume']
         self.assertEqual(fetched_volume['name'],
                          self.volume_name)
+        self.volumes_client.delete_volume(self.volume['id'])
diff --git a/cinder/tests/unit/image/fake.py b/cinder/tests/unit/image/fake.py
index f721726..7026858 100644
--- a/cinder/tests/unit/image/fake.py
+++ b/cinder/tests/unit/image/fake.py
@@ -23,6 +23,7 @@ import uuid
 
 from cinder import exception
 import cinder.image.glance
+from cinder.tests.unit import fake_constants
 
 
 class _FakeImageService(object):
@@ -34,7 +35,7 @@ class _FakeImageService(object):
         # So, make sure we've got one..
         timestamp = datetime.datetime(2011, 1, 1, 1, 2, 3)
 
-        image1 = {'id': '155d900f-4e14-4e4c-a73d-069cbf4541e6',
+        image1 = {'id': fake_constants.IMAGE_ID,
                   'name': 'fakeimage123456',
                   'created_at': timestamp,
                   'updated_at': timestamp,
diff --git a/cinder/volume/manager.py b/cinder/volume/manager.py
index 9870017..04fb0c2 100644
--- a/cinder/volume/manager.py
+++ b/cinder/volume/manager.py
@@ -85,6 +85,12 @@ from cinder.volume import rpcapi as volume_rpcapi
 from cinder.volume import utils as vol_utils
 from cinder.volume import volume_types
 
+if six.PY2:
+    from imp import reload
+    import sys
+    reload(sys)
+    sys.setdefaultencoding('utf-8')
+
 LOG = logging.getLogger(__name__)
 
 QUOTAS = quota.QUOTAS
diff --git a/etc/cinder/policy.json b/etc/cinder/policy.json
index 6a65122..3b32cb3 100644
--- a/etc/cinder/policy.json
+++ b/etc/cinder/policy.json
@@ -98,16 +98,16 @@
     "snapshot_extension:snapshot_unmanage": "rule:admin_api",
     "snapshot_extension:list_manageable": "rule:admin_api",
 
-    "consistencygroup:create" : "group:nobody",
-    "consistencygroup:delete": "group:nobody",
-    "consistencygroup:update": "group:nobody",
-    "consistencygroup:get": "group:nobody",
-    "consistencygroup:get_all": "group:nobody",
-
-    "consistencygroup:create_cgsnapshot" : "group:nobody",
-    "consistencygroup:delete_cgsnapshot": "group:nobody",
-    "consistencygroup:get_cgsnapshot": "group:nobody",
-    "consistencygroup:get_all_cgsnapshots": "group:nobody",
+    "consistencygroup:create" : "",
+    "consistencygroup:delete": "",
+    "consistencygroup:update": "",
+    "consistencygroup:get": "",
+    "consistencygroup:get_all": "",
+
+    "consistencygroup:create_cgsnapshot" : "",
+    "consistencygroup:delete_cgsnapshot": "",
+    "consistencygroup:get_cgsnapshot": "",
+    "consistencygroup:get_all_cgsnapshots": "",
 
     "group:group_types_manage": "rule:admin_api",
     "group:group_types_specs": "rule:admin_api",
diff --git a/test-requirements.txt b/test-requirements.txt
index 45b50a6..61dafbb 100644
--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -22,6 +22,5 @@ testscenarios>=0.4 # Apache-2.0/BSD
 oslosphinx>=4.7.0 # Apache-2.0
 oslo.versionedobjects[fixtures]>=1.17.0 # Apache-2.0
 os-testr>=0.8.0 # Apache-2.0
-tempest>=12.1.0 # Apache-2.0
 bandit>=1.1.0 # Apache-2.0
 reno>=1.8.0 # Apache-2.0
diff --git a/tox.ini b/tox.ini
index 61cda1a..8a5a18e 100644
--- a/tox.ini
+++ b/tox.ini
@@ -10,9 +10,10 @@ setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
          PYTHONWARNINGS=default::DeprecationWarning
 usedevelop = True
-install_command = pip install -c{env:UPPER_CONSTRAINTS_FILE:https://git.openstack.org/cgit/openstack/requirements/plain/upper-constraints.txt?h=stable/ocata} {opts} {packages}
+install_command = pip install -c{toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 
 # By default ostestr will set concurrency
 # to ncpu, to specify something else use
@@ -32,6 +33,7 @@ passenv = *_proxy *_PROXY
 # the API Ref to developer.openstack.org.
 whitelist_externals = rm
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
   rm -rf api-ref/build
   sphinx-build -W -b html -d api-ref/build/doctrees api-ref/source api-ref/build/html/
@@ -69,6 +71,7 @@ commands =
 
 [testenv:pylint]
 deps = -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        pylint==0.26.0
 commands = bash tools/lintstack.sh
 
@@ -107,6 +110,7 @@ commands = python {toxinidir}/tools/generate_driver_list.py
 
 [testenv:bandit]
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands = bandit -r cinder -n5 -x tests -ll
 
 [testenv:bindep]
