diff --git a/cinder/api/openstack/wsgi.py b/cinder/api/openstack/wsgi.py
index eeb9df1..96cda05 100644
--- a/cinder/api/openstack/wsgi.py
+++ b/cinder/api/openstack/wsgi.py
@@ -41,6 +41,10 @@ from cinder import policy
 from cinder import utils
 from cinder.wsgi import common as wsgi
 
+if not six.PY3:
+    import sys
+    reload(sys)
+    sys.setdefaultencoding('utf-8')
 
 LOG = logging.getLogger(__name__)
 
diff --git a/cinder/tests/tempest/api/volume/test_consistencygroups.py b/cinder/tests/tempest/api/volume/test_consistencygroups.py
index c524e9a..b7bcc54 100644
--- a/cinder/tests/tempest/api/volume/test_consistencygroups.py
+++ b/cinder/tests/tempest/api/volume/test_consistencygroups.py
@@ -14,25 +14,25 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
+from tempest.api.volume import base
 from tempest.common import waiters
 from tempest import config
 from tempest.lib.common.utils import data_utils
 from tempest.lib import decorators
 
-from cinder.tests.tempest.api.volume import base
 from cinder.tests.tempest import cinder_clients
 
 CONF = config.CONF
 
 
-class ConsistencyGroupsV2Test(base.BaseVolumeTest):
+class ConsistencyGroupsV2Test(base.BaseVolumeAdminTest):
     @classmethod
     def setup_clients(cls):
         cls._api_version = 2
         super(ConsistencyGroupsV2Test, cls).setup_clients()
         cls.admin_volume_client = cls.os_admin.volumes_v2_client
 
-        manager = cinder_clients.Manager(cls.os_adm)
+        manager = cinder_clients.Manager(cls.os_admin)
         cls.consistencygroups_adm_client = manager.consistencygroups_adm_client
 
     @classmethod
diff --git a/cinder/volume/manager.py b/cinder/volume/manager.py
index a6d40d6..7415af4 100644
--- a/cinder/volume/manager.py
+++ b/cinder/volume/manager.py
@@ -84,6 +84,11 @@ from cinder.volume import rpcapi as volume_rpcapi
 from cinder.volume import utils as vol_utils
 from cinder.volume import volume_types
 
+if not six.PY3:
+    import sys
+    reload(sys)
+    sys.setdefaultencoding('utf-8')
+
 LOG = logging.getLogger(__name__)
 
 QUOTAS = quota.QUOTAS
diff --git a/etc/cinder/policy.json b/etc/cinder/policy.json
index c51f564..5f6553b 100644
--- a/etc/cinder/policy.json
+++ b/etc/cinder/policy.json
@@ -114,16 +114,16 @@
     "snapshot_extension:snapshot_unmanage": "rule:admin_api",
     "snapshot_extension:list_manageable": "rule:admin_api",
 
-    "consistencygroup:create" : "group:nobody",
-    "consistencygroup:delete": "group:nobody",
-    "consistencygroup:update": "group:nobody",
-    "consistencygroup:get": "group:nobody",
-    "consistencygroup:get_all": "group:nobody",
-
-    "consistencygroup:create_cgsnapshot" : "group:nobody",
-    "consistencygroup:delete_cgsnapshot": "group:nobody",
-    "consistencygroup:get_cgsnapshot": "group:nobody",
-    "consistencygroup:get_all_cgsnapshots": "group:nobody",
+    "consistencygroup:create" : "",
+    "consistencygroup:delete": "",
+    "consistencygroup:update": "",
+    "consistencygroup:get": "",
+    "consistencygroup:get_all": "",
+
+    "consistencygroup:create_cgsnapshot" : "",
+    "consistencygroup:delete_cgsnapshot": "",
+    "consistencygroup:get_cgsnapshot": "",
+    "consistencygroup:get_all_cgsnapshots": "",
 
     "group:group_types_manage": "rule:admin_api",
     "group:group_types_specs": "rule:admin_api",
diff --git a/test-requirements.txt b/test-requirements.txt
index 8f25d59..764a865 100644
--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -21,7 +21,6 @@ testscenarios>=0.4 # Apache-2.0/BSD
 openstackdocstheme>=1.16.0 # Apache-2.0
 oslo.versionedobjects[fixtures]>=1.17.0 # Apache-2.0
 os-testr>=0.8.0 # Apache-2.0
-tempest>=16.1.0 # Apache-2.0
 bandit>=1.1.0 # Apache-2.0
 reno!=2.3.1,>=1.8.0 # Apache-2.0
 doc8 # Apache-2.0
diff --git a/tox.ini b/tox.ini
index bf4c656..ac435ed 100644
--- a/tox.ini
+++ b/tox.ini
@@ -10,10 +10,11 @@ setenv = VIRTUAL_ENV={envdir}
          PYTHONHASHSEED=0
          PYTHONWARNINGS=default::DeprecationWarning
 usedevelop = True
-install_command = pip install -c{env:UPPER_CONSTRAINTS_FILE:https://git.openstack.org/cgit/openstack/requirements/plain/upper-constraints.txt?h=stable/pike} {opts} {packages}
+install_command = pip install -c{toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 deps =
     -r{toxinidir}/test-requirements.txt
     -r{toxinidir}/requirements.txt
+    -c{toxinidir}/../requirements/upper-constraints.txt
 
 # By default ostestr will set concurrency
 # to ncpu, to specify something else use
@@ -21,7 +22,7 @@ deps =
 # call ie: 'tox -epy27 -- --concurrency=4'
 commands =
   find . -type f -name "*.pyc" -delete
-  ostestr {posargs}
+  ostestr --no-slowest --serial --regex '{posargs}'
 
 whitelist_externals =
   bash
@@ -33,6 +34,7 @@ passenv = *_proxy *_PROXY
 # the API Ref to developer.openstack.org.
 whitelist_externals = rm
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
   rm -rf api-ref/build
   sphinx-build -W -b html -d api-ref/build/doctrees api-ref/source api-ref/build/html/
@@ -68,6 +70,7 @@ commands =
 
 [testenv:pylint]
 deps = -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
        pylint==0.26.0
 commands = bash tools/lintstack.sh
 
@@ -107,6 +110,7 @@ commands = python {toxinidir}/tools/generate_driver_list.py
 
 [testenv:bandit]
 deps = -r{toxinidir}/test-requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands = bandit -r cinder -n5 -x tests -ll
 
 [testenv:bandit-baseline]
