diff --git a/tox.ini b/tox.ini
index edfee15..fdb814d 100644
--- a/tox.ini
+++ b/tox.ini
@@ -21,6 +21,21 @@ commands =
          find . -type f -name "*.pyc" -delete
          bash tools/pretty_tox.sh '{posargs}'
 
+[testenv:ci]
+envdir = .tox/ci
+sitepackages = {[tempestenv]sitepackages}
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+commands =
+    sh -c 'sudo chown -R $(id -un):$(id -gn) {toxinidir}'
+    find . -type f -name '*.pyc' -delete
+    ostestr --no-slowest --serial --regex '{posargs}'
+    sh -c 'test ! -e .testrepository -a -d .stestr && ln -s .stestr .testrepository || true'
+    sh -c 'testr last --subunit >report.sub'
+    sh -c 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file tools/config/config-generator.tempest.conf
 
diff --git a/.stestr.conf b/.stestr.conf
new file mode 100644
index 0000000..2cd3380
--- /dev/null
+++ b/.stestr.conf
@@ -0,0 +1,3 @@
+[DEFAULT]
+test_path=./tempest/test_discover
+group_regex=([^\.]*\.)*
