diff --git a/tox.ini b/tox.ini
index 44162fd..65f6440 100644
--- a/tox.ini
+++ b/tox.ini
@@ -11,6 +11,7 @@ setenv =
 deps =
     setuptools
     -r{toxinidir}/requirements.txt
+    -c{env:basedir}/requirements/upper-constraints.txt
 
 [testenv]
 setenv =
@@ -23,10 +24,28 @@ whitelist_externals = *
 deps =
     -r{toxinidir}/requirements.txt
     -r{toxinidir}/test-requirements.txt
+    -c{env:basedir}/requirements/upper-constraints.txt
 commands =
     find . -type f -name "*.pyc" -delete
     bash tools/pretty_tox.sh '{posargs}'
 
+[testenv:ci]
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+sitepackages = True
+recreate = True
+envdir = .tox/ci
+commands =
+    bash -cexu 'sudo chown -R $(id -un):$(id -gn) {toxinidir}'
+    find {toxinidir} -type f -name '*.pyc' -delete
+    bash -cexu 'mtp={env:basedir}/manila-tempest-plugin; if [[ -d $mtp ]]; then pip install $mtp; fi'
+    ostestr --no-slowest --serial --regex '{posargs}'
+    bash -cexu 'if [[ ! -e .testrepository && -d .stestr ]]; then ln -s .stestr .testrepository; fi'
+    bash -cexu 'testr last --subunit >report.sub'
+    bash -cexu 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file etc/config-generator.tempest.conf
 
diff --git a/.stestr.conf b/.stestr.conf
new file mode 100644
index 0000000..818c743
--- /dev/null
+++ b/.stestr.conf
@@ -0,0 +1,3 @@
+[DEFAULT]
+test_path=./tempest/test_discover
+group_regex=([^\.]*\.)*
