diff --git a/tox.ini b/tox.ini
index 9103175..69ddf42 100644
--- a/tox.ini
+++ b/tox.ini
@@ -32,6 +32,20 @@ commands =
     find . -type f -name "*.pyc" -delete
     stestr --test-path ./tempest/tests run {posargs}
 
+[testenv:ci]
+envdir = .tox/tempest
+sitepackages = {[tempestenv]sitepackages}
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+commands =
+    find . -type f -name '*.pyc' -delete
+    ostestr --no-slowest --serial --regex '{posargs}'
+    sh -c 'test ! -e .testrepository -a -d .stestr && ln -s .stestr .testrepository || true'
+    sh -c 'testr last --subunit >report.sub'
+    sh -c 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file tempest/cmd/config-generator.tempest.conf
 
