diff --git a/tempest/api/volume/admin/test_volume_retype.py b/tempest/api/volume/admin/test_volume_retype.py
index 18e0b9bff..c60b0da69 100644
--- a/tempest/api/volume/admin/test_volume_retype.py
+++ b/tempest/api/volume/admin/test_volume_retype.py
@@ -77,7 +77,6 @@ class VolumeRetypeTest(base.BaseVolumeAdminTest):
             volume['id'],
             new_type=self.dst_vol_type['name'],
             migration_policy=migration_policy)
-        self.addCleanup(self._wait_for_internal_volume_cleanup, volume)
         waiters.wait_for_volume_retype(self.volumes_client, volume['id'],
                                        self.dst_vol_type['name'])
 
diff --git a/tox.ini b/tox.ini
index e861c8482..fe851c845 100644
--- a/tox.ini
+++ b/tox.ini
@@ -11,7 +11,7 @@ setenv =
     VIRTUAL_ENV={envdir}
     OS_TEST_PATH=./tempest/test_discover
 deps =
-    -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+    -c{toxinidir}/../requirements/upper-constraints.txt
     -r{toxinidir}/requirements.txt
 
 [testenv]
@@ -25,16 +25,33 @@ setenv =
     PYTHONWARNINGS=default::DeprecationWarning,ignore::DeprecationWarning:distutils,ignore::DeprecationWarning:site
 passenv = OS_STDOUT_CAPTURE OS_STDERR_CAPTURE OS_TEST_TIMEOUT OS_TEST_LOCK_PATH TEMPEST_CONFIG TEMPEST_CONFIG_DIR http_proxy HTTP_PROXY https_proxy HTTPS_PROXY no_proxy NO_PROXY ZUUL_CACHE_DIR REQUIREMENTS_PIP_LOCATION GENERATE_TEMPEST_PLUGIN_LIST GABBI_TEMPEST_PATH
 usedevelop = True
-install_command = pip install {opts} {packages}
+install_command = pip install {opts} -c{toxinidir}/../requirements/upper-constraints.txt {packages}
 whitelist_externals = *
 deps =
-    -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+    -c{toxinidir}/../requirements/upper-constraints.txt
     -r{toxinidir}/requirements.txt
     -r{toxinidir}/test-requirements.txt
 commands =
     find . -type f -name "*.pyc" -delete
     stestr --test-path ./tempest/tests run {posargs}
 
+[testenv:ci]
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+sitepackages = True
+recreate = True
+envdir = .tox/ci
+commands =
+    bash -cexu 'sudo chown -R $(id -un):$(id -gn) {toxinidir}'
+    find {toxinidir} -type f -name '*.pyc' -delete
+    bash -cexu 'find {toxinidir}/.. -mindepth 1 -maxdepth 1 -type d -name *-tempest-plugin | xargs -L1 -rt pip install'
+    ostestr --no-slowest --serial --concurrency 0 --regex '{posargs}'
+    bash -cexu 'if [[ ! -e .testrepository && -d .stestr ]]; then ln -s .stestr .testrepository; fi'
+    bash -cexu 'stestr last --subunit >report.sub'
+    bash -cexu 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file tempest/cmd/config-generator.tempest.conf
 
@@ -263,7 +280,7 @@ commands =
 
 [testenv:venv]
 deps =
-  -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+  -c{toxinidir}/../requirements/upper-constraints.txt
   -r{toxinidir}/requirements.txt
   -r{toxinidir}/doc/requirements.txt
 commands = {posargs}
@@ -278,7 +295,7 @@ commands = {posargs}
 
 [testenv:docs]
 deps =
-  -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+  -c{toxinidir}/../requirements/upper-constraints.txt
   -r{toxinidir}/requirements.txt
   -r{toxinidir}/doc/requirements.txt
 commands =
@@ -296,6 +313,7 @@ commands =
 
 [testenv:pep8]
 deps =
+    -c{toxinidir}/../requirements/upper-constraints.txt
     -r{toxinidir}/test-requirements.txt
     autopep8
 commands =
@@ -348,7 +366,7 @@ paths =
 
 [testenv:releasenotes]
 deps =
-  -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
+  -c{toxinidir}/../requirements/upper-constraints.txt
   -r{toxinidir}/requirements.txt
   -r{toxinidir}/doc/requirements.txt
 commands =
@@ -376,6 +394,7 @@ commands = bash -c "find {toxinidir}/tools    \
 # This also means that pip-check-reqs must be installed separately, outside
 # of the requirements.txt files
 deps = pip_check_reqs
+       -c{toxinidir}/../requirements/upper-constraints.txt
        -r{toxinidir}/requirements.txt
 commands=
     pip-extra-reqs -d --ignore-file=tempest/tests/* tempest
