diff --git a/test-requirements.txt b/test-requirements.txt
index 2d15daa44..08d3d3ce6 100644
--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -1,7 +1,7 @@
 # The order of packages is significant, because pip processes them in the order
 # of appearance. Changing the order has an impact on the overall integration
 # process, which may cause wedges in the gate later.
-hacking<0.12,>=0.11.0 # Apache-2.0
+hacking<0.11,>=0.10.0 # Apache-2.0
 # needed for doc build
 sphinx!=1.3b1,<1.4,>=1.2.1 # BSD
 python-subunit>=0.0.18 # Apache-2.0/BSD
diff --git a/tox.ini b/tox.ini
index 02eef78f2..a8881f7e0 100644
--- a/tox.ini
+++ b/tox.ini
@@ -11,6 +11,7 @@ setenv =
 deps =
     setuptools
     -r{toxinidir}/requirements.txt
+    -c{toxinidir}/../requirements/upper-constraints.txt
 
 [testenv]
 setenv =
@@ -19,15 +20,33 @@ setenv =
     PYTHONWARNINGS=default::DeprecationWarning
 passenv = OS_STDOUT_CAPTURE OS_STDERR_CAPTURE OS_TEST_TIMEOUT OS_TEST_LOCK_PATH OS_TEST_PATH TEMPEST_CONFIG TEMPEST_CONFIG_DIR http_proxy HTTP_PROXY https_proxy HTTPS_PROXY no_proxy NO_PROXY
 usedevelop = True
-install_command = pip install -U {opts} {packages}
+install_command = pip install {opts} -c{toxinidir}/../requirements/upper-constraints.txt {packages}
 whitelist_externals = *
 deps =
     -r{toxinidir}/requirements.txt
     -r{toxinidir}/test-requirements.txt
+    -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
     find . -type f -name "*.pyc" -delete
     ostestr {posargs}
 
+[testenv:ci]
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+sitepackages = True
+recreate = True
+envdir = .tox/ci
+commands =
+    bash -cexu 'sudo chown -R $(id -un):$(id -gn) {toxinidir}'
+    find {toxinidir} -type f -name '*.pyc' -delete
+    bash -cexu 'find {toxinidir}/.. -mindepth 1 -maxdepth 1 -type d -name *-tempest-plugin | xargs -L1 -rt pip install'
+    ostestr --no-slowest --serial --regex '{posargs}'
+    bash -cexu 'if [[ ! -e .testrepository && -d .stestr ]]; then ln -s .stestr .testrepository; fi'
+    bash -cexu 'testr last --subunit >report.sub'
+    bash -cexu 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file tempest/cmd/config-generator.tempest.conf
 
@@ -166,6 +185,7 @@ commands = sphinx-build -a -E -W -d releasenotes/build/doctrees -b html releasen
 # of the requirements.txt files
 deps = pip_check_reqs
        -r{toxinidir}/requirements.txt
+       -c{toxinidir}/../requirements/upper-constraints.txt
 commands=
     pip-extra-reqs -d --ignore-file=tempest/tests/* tempest
     pip-missing-reqs -d --ignore-file=tempest/tests/* tempest
diff --git a/.stestr.conf b/.stestr.conf
new file mode 100644
index 000000000..818c7436a
--- /dev/null
+++ b/.stestr.conf
@@ -0,0 +1,3 @@
+[DEFAULT]
+test_path=./tempest/test_discover
+group_regex=([^\.]*\.)*
