diff --git a/test-requirements.txt b/test-requirements.txt
index 13950bd..5811d6f 100644
--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -1,7 +1,7 @@
 # The order of packages is significant, because pip processes them in the order
 # of appearance. Changing the order has an impact on the overall integration
 # process, which may cause wedges in the gate later.
-hacking!=0.13.0,<0.14,>=0.12.0 # Apache-2.0
+hacking<0.11,>=0.10.0
 # needed for doc build
 sphinx>=1.5.1 # BSD
 oslosphinx>=4.7.0 # Apache-2.0
diff --git a/tox.ini b/tox.ini
index 892f834..a8e077c 100644
--- a/tox.ini
+++ b/tox.ini
@@ -12,6 +12,7 @@ setenv =
     CLIENT_NAME=tempest
 deps =
     -r{toxinidir}/requirements.txt
+    -c{toxinidir}/../requirements/upper-constraints.txt
 
 [testenv]
 setenv =
@@ -23,15 +24,33 @@ setenv =
 passenv = OS_STDOUT_CAPTURE OS_STDERR_CAPTURE OS_TEST_TIMEOUT OS_TEST_LOCK_PATH OS_TEST_PATH TEMPEST_CONFIG TEMPEST_CONFIG_DIR http_proxy HTTP_PROXY https_proxy HTTPS_PROXY no_proxy NO_PROXY ZUUL_CACHE_DIR REQUIREMENTS_PIP_LOCATION
 usedevelop = True
 install_command =
-    {toxinidir}/tools/tox_install.sh {env:UPPER_CONSTRAINTS_FILE:https://git.openstack.org/cgit/openstack/requirements/plain/upper-constraints.txt} {opts} {packages}
+    {toxinidir}/tools/tox_install.sh {toxinidir}/../requirements/upper-constraints.txt {opts} {packages}
 whitelist_externals = *
 deps =
     -r{toxinidir}/requirements.txt
     -r{toxinidir}/test-requirements.txt
+    -c{toxinidir}/../requirements/upper-constraints.txt
 commands =
     find . -type f -name "*.pyc" -delete
     ostestr {posargs}
 
+[testenv:ci]
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+sitepackages = True
+recreate = True
+envdir = .tox/ci
+commands =
+    bash -cexu 'sudo chown -R $(id -un):$(id -gn) {toxinidir}'
+    find {toxinidir} -type f -name '*.pyc' -delete
+    bash -cexu 'find {toxinidir}/.. -mindepth 1 -maxdepth 1 -type d -name *-tempest-plugin | xargs -L1 -rt pip install'
+    ostestr --no-slowest --serial --regex '{posargs}'
+    bash -cexu 'if [[ ! -e .testrepository && -d .stestr ]]; then ln -s .stestr .testrepository; fi'
+    bash -cexu 'testr last --subunit >report.sub'
+    bash -cexu 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file tempest/cmd/config-generator.tempest.conf
 
diff --git a/.stestr.conf b/.stestr.conf
new file mode 100644
index 0000000..818c743
--- /dev/null
+++ b/.stestr.conf
@@ -0,0 +1,3 @@
+[DEFAULT]
+test_path=./tempest/test_discover
+group_regex=([^\.]*\.)*
