diff --git a/tox.ini b/tox.ini
index effd400..98ed9cc 100644
--- a/tox.ini
+++ b/tox.ini
@@ -22,7 +22,7 @@ setenv =
     PYTHONWARNINGS=default::DeprecationWarning,ignore::DeprecationWarning:distutils,ignore::DeprecationWarning:site
 passenv = OS_STDOUT_CAPTURE OS_STDERR_CAPTURE OS_TEST_TIMEOUT OS_TEST_LOCK_PATH TEMPEST_CONFIG TEMPEST_CONFIG_DIR http_proxy HTTP_PROXY https_proxy HTTPS_PROXY no_proxy NO_PROXY ZUUL_CACHE_DIR REQUIREMENTS_PIP_LOCATION GENERATE_TEMPEST_PLUGIN_LIST GABBI_TEMPEST_PATH
 usedevelop = True
-install_command = pip install {opts} {packages}
+install_command = pip install {opts} -c{env:UPPER_CONSTRAINTS_FILE} {packages}
 whitelist_externals = *
 deps =
     -c{env:UPPER_CONSTRAINTS_FILE:https://releases.openstack.org/constraints/upper/master}
@@ -32,6 +32,23 @@ commands =
     find . -type f -name "*.pyc" -delete
     stestr --test-path ./tempest/tests run {posargs}
 
+[testenv:ci]
+setenv = {[tempestenv]setenv}
+deps = {[tempestenv]deps}
+ignore_errors = True
+sitepackages = True
+recreate = True
+envdir = .tox/ci
+commands =
+    bash -cexu 'sudo chown -R $(id -un):$(id -gn) {toxinidir}'
+    find {toxinidir} -type f -name '*.pyc' -delete
+    bash -cexu 'find {toxinidir}/.. -mindepth 1 -maxdepth 1 -type d -name *-tempest-plugin | xargs -L1 -rt pip install'
+    ostestr --no-slowest --serial --concurrency 0 --regex '{posargs}'
+    bash -cexu 'if [[ ! -e .testrepository && -d .stestr ]]; then ln -s .stestr .testrepository; fi'
+    bash -cexu 'stestr last --subunit >report.sub'
+    bash -cexu 'cat report.sub | subunit-trace >report.txt'
+    subunit2html report.sub report.html
+
 [testenv:genconfig]
 commands = oslo-config-generator --config-file tempest/cmd/config-generator.tempest.conf
 
