diff --git a/manila/api/v2/shares.py b/manila/api/v2/shares.py
index 13bb08f..ac7c519 100644
--- a/manila/api/v2/shares.py
+++ b/manila/api/v2/shares.py
@@ -149,13 +149,13 @@ class ShareController(shares.ShareMixin,
 
             self.share_api.revert_to_snapshot(context, share, snapshot)
         except exception.ShareNotFound as e:
-            raise exc.HTTPNotFound(explanation=e)
+            raise exc.HTTPNotFound(explanation=e.msg)
         except exception.ShareSnapshotNotFound as e:
-            raise exc.HTTPBadRequest(explanation=e)
+            raise exc.HTTPBadRequest(explanation=e.msg)
         except exception.ShareSizeExceedsAvailableQuota as e:
-            raise exc.HTTPForbidden(explanation=e)
+            raise exc.HTTPForbidden(explanation=e.msg)
         except exception.ReplicationException as e:
-            raise exc.HTTPBadRequest(explanation=e)
+            raise exc.HTTPBadRequest(explanation=e.msg)
 
         return webob.Response(status_int=http_client.ACCEPTED)
 
